File: clients/typescript/src/jobs.ts
// jobs.ts provides waitForJob, a poll helper for Crucible async jobs. It
// complements the generated client in client.ts (same package) —
// hand-maintained, NOT written by scripts/gen-clients.sh, exactly like
// webhook.ts.
import { ApiError } from "./errors";
import type { Client, GetJobResponse } from "./client";

/** Terminal status values returned by getJob. Mirror the gateway's
 * asyncJobResponse.status wire contract (gateway/internal/server/routes.go) —
 * part of the frozen contract, changes only alongside it. */
export const JOB_STATUS_SUCCEEDED = "succeeded";
export const JOB_STATUS_FAILED = "failed";

/** Default delay between getJob polls, used when WaitForJobOptions.pollIntervalMs is omitted. */
export const DEFAULT_POLL_INTERVAL_MS = 1000;

/** Thrown by waitForJob when the caller's signal aborts or timeoutMs elapses before the job reaches a terminal status. */
export class JobWaitAbortedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "JobWaitAbortedError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export interface WaitForJobOptions {
  /** Delay between getJob polls, in ms. Defaults to DEFAULT_POLL_INTERVAL_MS. */
  pollIntervalMs?: number;
  /** Total time budget for the wait, in ms. Omit for no additional timeout. */
  timeoutMs?: number;
  /** Caller-supplied cancellation signal; aborting it stops polling immediately. */
  signal?: AbortSignal;
  /** Override the default API key for the underlying getJob calls. */
  apiKey?: string;
}

// mergeAbortSignals returns a signal that aborts as soon as any input signal
// aborts. Implemented manually (not AbortSignal.any) because that API landed
// in Node 20; this SDK's engines floor is Node >=18 (package.json).
function mergeAbortSignals(signals: AbortSignal[]): AbortSignal {
  const controller = new AbortController();
  for (const s of signals) {
    if (s.aborted) {
      controller.abort(s.reason);
      break;
    }
    s.addEventListener("abort", () => controller.abort(s.reason), { once: true });
  }
  return controller.signal;
}

function sleep(ms: number, signal: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal.aborted) {
      reject(signal.reason);
      return;
    }
    const onAbort = () => {
      clearTimeout(timer);
      reject(signal.reason);
    };
    const timer = setTimeout(() => {
      signal.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    signal.addEventListener("abort", onAbort, { once: true });
  });
}

// jobErrorToApiError maps GetJobResponse.error (the gateway's asyncJobError
// JSON shape: {"code":"...","message":"..."}) to the SDK's typed ApiError, so
// a failed job surfaces through the same error type as an HTTP-level failure
// rather than a raw status string. Falls back to a generic description if the
// field is missing or shaped unexpectedly.
function jobErrorToApiError(raw: unknown): ApiError {
  const obj = (raw && typeof raw === "object" ? raw : {}) as { code?: unknown; message?: unknown };
  const code = typeof obj.code === "string" && obj.code ? obj.code : "UNKNOWN";
  const message = typeof obj.message === "string" && obj.message ? obj.message : "job failed";
  return new ApiError(0, code, message);
}

/**
 * waitForJob polls client.getJob(jobId) until the job reaches a terminal
 * status, options.signal aborts, or options.timeoutMs elapses — whichever
 * comes first. On "succeeded" it resolves with the job's final
 * GetJobResponse. On "failed" it rejects with an ApiError built from the
 * job's recorded error code/message. No new HTTP route is introduced: every
 * poll is a plain getJob call.
 */
export async function waitForJob(
  client: Client,
  jobId: string,
  options: WaitForJobOptions = {},
): Promise<GetJobResponse> {
  const pollIntervalMs = options.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;

  const signals: AbortSignal[] = [];
  if (options.signal) signals.push(options.signal);
  let timeoutTimer: ReturnType<typeof setTimeout> | undefined;
  if (options.timeoutMs !== undefined) {
    const timeoutController = new AbortController();
    timeoutTimer = setTimeout(
      () => timeoutController.abort(new JobWaitAbortedError(`waitForJob: timed out after ${options.timeoutMs}ms`)),
      options.timeoutMs,
    );
    signals.push(timeoutController.signal);
  }
  const combined = signals.length > 0 ? mergeAbortSignals(signals) : new AbortController().signal;

  try {
    for (;;) {
      if (combined.aborted) {
        throw combined.reason ?? new JobWaitAbortedError("waitForJob: aborted");
      }
      const job = await client.getJob(jobId, options.apiKey);
      if (job.status === JOB_STATUS_SUCCEEDED) {
        return job;
      }
      if (job.status === JOB_STATUS_FAILED) {
        throw jobErrorToApiError(job.error);
      }
      await sleep(pollIntervalMs, combined);
    }
  } finally {
    if (timeoutTimer !== undefined) clearTimeout(timeoutTimer);
  }
}

File: clients/go/jobs.go
// jobs.go provides WaitForJob, a poll helper for Crucible async jobs. It
// complements the generated client in client.go (same package, crucible) —
// hand-maintained, NOT written by scripts/gen-clients.sh, exactly like
// webhook.go.
package crucible

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

// Terminal status values returned by GetJob. Mirror gateway/internal/jobs.Job's
// Status constants (StatusSucceeded, StatusFailed, StatusCancelled) — the
// client package cannot import the gateway's internal package, so the
// strings are duplicated here; they are part of the frozen wire contract
// (asyncJobResponse.Status in gateway/internal/server/routes.go) and change
// only alongside it.
const (
	JobStatusSucceeded = "succeeded"
	JobStatusFailed    = "failed"
	JobStatusCancelled = "cancelled"
)

// CancelJob calls POST /v1/jobs/{id}/cancel: withdraws apiKey's own
// still-queued job. Hand-maintained here rather than generated into
// client.go: scripts/gen-clients.sh emits one method per operationId
// declared in clients/openapi.json, but this file already hosts the SDK's
// other hand-maintained job helpers (WaitForJob below), and CancelJob's
// response is byte-identical in shape to GetJob's (the gateway's
// jobsCancelHandler returns the same asyncJobResponse envelope
// jobsGetHandler does), so it reuses GetJobResponse rather than
// introducing a duplicate type. Uses the same c.do/checkError primitives
// client.go's generated methods use — same package, so both are visible
// here.
func (c *Client) CancelJob(ctx context.Context, apiKey, jobID string) (*GetJobResponse, error) {
	path := fmt.Sprintf("/v1/jobs/%s/cancel", url.PathEscape(jobID))
	resp, err := c.do(ctx, http.MethodPost, path, apiKey, nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := checkError(resp); err != nil {
		return nil, err
	}
	var out GetJobResponse
	if decErr := json.NewDecoder(resp.Body).Decode(&out); decErr != nil {
		return nil, fmt.Errorf("crucible: decode response: %w", decErr)
	}
	return &out, nil
}

// DefaultPollInterval is used by WaitForJob when WaitForJobOptions is nil or
// its PollInterval is zero.
const DefaultPollInterval = 1 * time.Second

// WaitForJobOptions configures WaitForJob. The zero value is valid: it polls
// every DefaultPollInterval with no additional timeout beyond ctx.
type WaitForJobOptions struct {
	// PollInterval is the delay between GetJob calls. Zero means DefaultPollInterval.
	PollInterval time.Duration
	// Timeout bounds the total wait, in addition to whatever deadline ctx
	// already carries. Zero means no additional timeout is applied.
	Timeout time.Duration
}

// JobFailedError is returned by WaitForJob when the job reaches the "failed"
// terminal status. It wraps the same code/message the gateway recorded on the
// job row (SanitizeWorkerError-filtered — see gateway/internal/jobs.Job), so
// callers get a typed error rather than a raw status string.
type JobFailedError struct {
	Code    string
	Message string
}

func (e *JobFailedError) Error() string {
	return "crucible: job failed: " + e.Code + ": " + e.Message
}

// WaitForJob polls GetJob(ctx, apiKey, jobID) until the job reaches a terminal
// status, ctx is cancelled/expires, or opts.Timeout elapses — whichever comes
// first. On "succeeded" or "cancelled" it returns the job's final
// GetJobResponse (callers distinguish the two via Status); on "failed" it
// returns a *JobFailedError built from the job's recorded error code/message.
// No new HTTP route is introduced: every poll is a plain GetJob call.
func (c *Client) WaitForJob(ctx context.Context, apiKey, jobID string, opts *WaitForJobOptions) (*GetJobResponse, error) {
	interval := DefaultPollInterval
	var timeout time.Duration
	if opts != nil {
		if opts.PollInterval > 0 {
			interval = opts.PollInterval
		}
		timeout = opts.Timeout
	}
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}

	for {
		job, err := c.GetJob(ctx, apiKey, jobID)
		if err != nil {
			return nil, err
		}
		switch job.Status {
		case JobStatusSucceeded, JobStatusCancelled:
			return job, nil
		case JobStatusFailed:
			code, message := jobErrorDetails(job.Error)
			return nil, &JobFailedError{Code: code, Message: message}
		}

		timer := time.NewTimer(interval)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, ctx.Err()
		case <-timer.C:
		}
	}
}

// jobErrorDetails extracts code/message from GetJobResponse.Error, which
// decodes as map[string]any (the gateway's asyncJobError JSON shape:
// {"code":"...","message":"..."}). Falls back to a generic description if the
// field is missing or shaped unexpectedly, so WaitForJob never panics on a
// malformed/absent error payload.
func jobErrorDetails(raw any) (code, message string) {
	code, message = "UNKNOWN", "job failed"
	m, ok := raw.(map[string]any)
	if !ok {
		return code, message
	}
	if c, ok := m["code"].(string); ok && c != "" {
		code = c
	}
	if msg, ok := m["message"].(string); ok && msg != "" {
		message = msg
	}
	return code, message
}
